src/edu/washington/escience/myria/coordinator/catalog/MasterCatalog.java

```diff
@@ -1753,6 +1753,40 @@ private void deleteRelationIfExists(@Nonnull final SQLiteConnection sqliteConnec
   }
 
   /**
+   * Connects to SQLite to delete
+   * 
+   * @throws CatalogException if there is an error in the catalog.
+   */
+  public void deleteRelationFromCatalog(final RelationKey relationkey) throws CatalogException {
+    if (isClosed) {
+      throw new CatalogException("Catalog is closed.");
+    }
+    /* Do the work */
+    try {
+      queue.execute(new SQLiteJob<Void>() {
+        @Override
+        protected Void job(final SQLiteConnection sqliteConnection) throws CatalogException, SQLiteException {
+          sqliteConnection.exec("BEGIN TRANSACTION;");
+          try {
+            deleteRelationIfExists(sqliteConnection, relationkey);
+            sqliteConnection.exec("COMMIT TRANSACTION;");
+          } catch (SQLiteException e) {
+            try {
+              sqliteConnection.exec("ROLLBACK TRANSACTION;");
+            } catch (SQLiteException e2) {
+              assert true; /* Do nothing */
+            }
+            throw e;
+          }
+          return null;
+        }
+      }).get();
+    } catch (InterruptedException | ExecutionException e) {
```

Comment:
Actually, now that I think about it, only the visibility of the deleted relation needs to be atomic, and that can be accomplished just with a transactional update of the master catalog. The data deletion on the workers doesn't need to be transactional at all, and could simply retry until it succeeds. In fact, it could be asynchronous: the workers could periodically compare the local Postgres catalog to the master catalog and delete any tables corresponding to relations not present in the master catalog. Of course, we'd need to anticipate possible race conditions--perhaps a Postgres advisory lock that would need to be acquired by workers to either create local partitions or delete them would work. These are just some half-baked ideas--I guess we should discuss the design in the next Myria meeting.